OSSM-8215 [DOC] Document using multiple control planes on a single cluster (#85272)
openshift/openshift-docs: modules/ossm-deploy-application-workloads-per-user-group.adoc

// Module included in the following assemblies:
// install/ossm-using-multiple-control-planes-on-single-cluster.adoc

:_mod-docs-content-type: Procedure
[id="ossm-deploy-application-workloads-per-user-group_{context}"]

> = Deploy application workloads per user group

Reviewer: ```suggestion
= Deploy application workloads in each mesh
```
The term "user group" seems very specific to the example (taken from the [upstream Istio doc](https://istio.io/latest/docs/setup/install/multiple-controlplanes/)), but it is just an example and not an actual Istio concept, so I think saying "Deploy application workloads in each mesh" better describes the scenario.

Reviewer 2: @longmuir Why even use the term "user group"? This document describes deploying two meshes on the same cluster, so why introduce an additional concept (and labels) when the point is simply to mark the mesh each namespace belongs to?

Reviewer: Agreed. It was from the upstream Istio doc, and I removed a lot of its use here because it felt confusing, but I left it for the discoverySelector labels. I see you just used `mesh=mesh1` in your example doc for discovery selectors. I think you're right that it's good remove the term completely to avoid confusion that it is an Istio concept, rather than just a choice of naming.

I also like how you broke out the discovery selector labelling step up from the injection labelling, and I think that's a good idea too. 

I'll suggest changes for both - thanks!

:context: ossm-deploy-application-workloads-per-user-group

This procedure demonstrates how to deploy application workloads in separate namespaces based on user groups.

.Procedure

. Create an application namespace called `app-ns-1`:
+
[source,terminal]
----
$ oc create namespace app-ns-1
----

> . Label the namespace with the respective control plane:

Reviewer: ```suggestion
. Add the `usergroup=usergroup-1` label so that the namespace is discovered by the first control plane and the `istio.io/rev=mesh-1` label so that side car injection is enabled in the namespace by default. Note that `mesh-1` corresponds to the revision name, which can be confirmed with the command `oc get istiorevisions`:
```

+
[source,terminal]
----

> $ oc label namespace app-ns-1 usergroup=usergroup-1 istio.io/rev=usergroup-1

Reviewer: ```suggestion
$ oc label namespace app-ns-1 usergroup=usergroup-1 istio.io/rev=mesh-1
```

----

. Deploy the `sleep` and `httpbin` applications:
+
[source,terminal]
----

> $ oc apply -f https://raw.githubusercontent.com/istio/istio/release-1.23/samples/sleep/sleep.yaml -n app-ns-1

Reviewer: @FilipB Similar issue here with referencing upstream artifacts. If we want to use these, they should be from our downstream "openshift-service-mesh" fork. 

For httpbin, we should also reference a release branch rather than master branch in case it gets removed in a future release.

Would the following work instead?
https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml
https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml

Reviewer 3: IIRC the version of httpbin from release-1.23 was not working so it was necessary to use the version from master. It seems it's [fixed](https://github.com/istio/istio/commit/6649dd8ed6b6cb58a3e04598fa72a25cf539d58b) now even in release-1.23 so it's safe to use it.
Should I sync changes from here to openshift-service-mesh/sail-operator repo?

> $ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml

Reviewer 4: $ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml -n app-ns-1

> $ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml

Reviewer 4: $ oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml -n app-ns-1

----

. Wait for the `httpbin` and `sleep` pods to run with sidecars injected:
+
[source,terminal]
----
$ oc get pods -n app-ns-1
----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
NAME                       READY   STATUS    RESTARTS   AGE
httpbin-7f56dc944b-kpw2x   2/2     Running   0          2m26s
sleep-5577c64d7c-b5wd2     2/2     Running   0          91m
----

> . Repeat the above steps for two more namespaces: `app-ns-2` and `app-ns-3`.

Reviewer: ```suggestion
. Repeat the above steps for two more namespaces: `app-ns-2` and `app-ns-3`, adding them both to the `usergroup-2` revision.
```
It wasn't clear where these should be created otherwise.

> . Repeat the above steps for two more namespaces: `app-ns-2` and `app-ns-3`, adding them both to the `usergroup-2` revision.

Reviewer: I think we should provide the steps for this explicitly, as the label command is not necessarily that obvious and it's easy to put the wrong thing in the wrong place. I hope it is ok to put multiple commands in some boxes, especially when they are identical and we have already talked about them above. The content would be:

---
1. Create two more namespaces: `app-ns-2` and `app-ns-3`:
```
oc create namespace app-ns-2
oc create namespace app-ns-3
```
2.  Add them both to the second control plane with the discovery selector `usergroup=usergroup-2` and the revision label with the value `mesh-2`:
```
oc label namespace app-ns-2 usergroup=usergroup-2 istio.io/rev=mesh-2
oc label namespace app-ns-3 usergroup=usergroup-2 istio.io/rev=mesh-2
```
3. Deploy the `sleep` and `httbin` applications to both namespaces:
```
oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml -n app-ns-2
oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml -n app-ns-2
oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml -n app-ns-3
oc apply -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml -n app-ns-3
```

Author: Placing the commands in separate boxes as per doc guidelines.

Reviewer 2: Note: you can convert this into two commands, so you'll only need two boxes instead of four (and the reader will only need to copy-paste 2x instead of 4x):
```
oc apply -n app-ns-2 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml

oc apply -n app-ns-3 \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/sleep/sleep.yaml \
   -f https://raw.githubusercontent.com/openshift-service-mesh/istio/release-1.23/samples/httpbin/httpbin.yaml
```

[NOTE]
====
You can use the following command to wait for deployment to be ready:
[source,terminal]
----
$ oc wait deployment/sleep -n app-ns-1 --for=condition=Available
----
====

.Verification

> After deploying the applications, use the `istioctl ps` command to verify that each application workload is managed by its assigned control plane. For example, `app-ns-1` is managed by `usergroup-1`, and `app-ns-2` and `app-ns-3` are managed by `usergroup-2`.

Reviewer: ```suggestion
After deploying the applications, use the `istioctl ps` command to verify that each application workload is managed by its assigned control plane. For example, `app-ns-1` is managed by `istio-system-1`, and `app-ns-2` and `app-ns-3` are managed by `istio-system-2`.
```

> After deploying the applications, use the `istioctl ps` command to verify that each application workload is managed by its assigned control plane. For example, `app-ns-1` is managed by `mesh-1`, and `app-ns-2` and `app-ns-3` are managed by `mesh-2`.

Reviewer: I found this step really useful, but also not easy to understand, because "istioctl ps" on its own doesn't work, and you need to specify the namespace of the control plane, which isn't given in this paragraph. It could also use example output. 

I suggest the following content replace this paragraph (@shreyasiddhartha feel free to adjust the formatting):

----
After deploying the applications, use the `istioctl ps` command to verify that each application workload is managed by its assigned control plane. 

For the first control plane deployed in `istio-system-1`:
```
istioctl ps -i istio-system-1
NAME                                  CLUSTER        CDS              LDS              EDS              RDS              ECDS        ISTIOD                            VERSION
httpbin-7f56dc944b-vwfm5.app-ns-1     Kubernetes     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     IGNORED     istiod-mesh-1-b69646b6f-kxrwk     1.23.0
sleep-5577c64d7c-d675f.app-ns-1       Kubernetes     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     SYNCED (11m)     IGNORED     istiod-mesh-1-b69646b6f-kxrwk     1.23.0
```
For the second control plane deployed in `istio-system-2`:
```
istioctl ps -i istio-system-2
NAME                                  CLUSTER        CDS                LDS                EDS                RDS                ECDS        ISTIOD                            VERSION
httpbin-7f56dc944b-54gjs.app-ns-3     Kubernetes     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     SYNCED (3m59s)     IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
httpbin-7f56dc944b-gnh72.app-ns-2     Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-k9mxz.app-ns-2       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
sleep-5577c64d7c-m9hvm.app-ns-3       Kubernetes     SYNCED (4m1s)      SYNCED (4m1s)      SYNCED (3m59s)     SYNCED (4m1s)      IGNORED     istiod-mesh-2-8666fdfc6-mqp45     1.23.0
```

Use this procedure to verify that application connectivity is restricted to workloads within their respective user groups:

> . Send a request from the `sleep` pod in `app-ns-1` in `usergroup-1` to the `httpbin` service in `app-ns-2` in `usergroup-2`. The communication should fail:

Reviewer: ```suggestion
. Send a request from the `sleep` pod in `app-ns-1` to the `httpbin` service in `app-ns-2`. The communication should fail:
```

+
[source,terminal]
----
$ oc -n app-ns-1 exec "$(oc -n app-ns-1 get pod -l app=sleep -o jsonpath={.items..metadata.name})" -c sleep -- curl -sIL http://httpbin.app-ns-2.svc.cluster.local:8000
----
+
You should see output similar to the following example:
+
.Example output
[source,terminal]
----
HTTP/1.1 503 Service Unavailable
content-length: 95
content-type: text/plain
date: Wed, 16 Oct 2024 12:05:37 GMT
server: envoy
----

> . Send a request from the `sleep` pod in `app-ns-2` in `usergroup-2` to the `httpbin` service in `app-ns-3` in `usergroup-2`. The communication should work:

Reviewer: ```suggestion
. Send a request from the `sleep` pod in `app-ns-2` managed by `mesh-2` to the `httpbin` service in `app-ns-3` also managed by `mesh-2`. The communication should work:
```
